File: apps/web/src/service/api/research-compute.ts
import { request } from "../request"

export interface ComputeResourceLimits {
  cpu_millis: number
  memory_mb: number
  gpu_count: number
  timeout_seconds: number
  max_output_bytes: number
}

export interface ResearchComputeEnvironment {
  key: string
  version: string
  kind: "compute"
  name: string
  description: string
  source_type: "research_compute_environment_revision"
  source_id: string
  source_revision_id: string
  executor_types: ["compute_runner"]
  risk: "low" | "medium" | "high"
  input_schema: Record<string, unknown>
  output_schema: Record<string, unknown>
  available: boolean
  unavailable_reason: string
  metadata: {
    lab_id: string
    environment_key: string
    environment_revision: number
    runner_protocol_version: "airalogy.compute-runner.v1"
    image_ref: string
    runtime_version: string
    allowed_languages: Array<"python" | "r">
    resource_limits: ComputeResourceLimits
    network_policy: "none" | "egress_allowlist"
    allowed_egress_hosts: string[]
    software_manifest: Record<string, unknown>
    estimated_cost_per_hour?: string | null
    currency?: string | null
    change_reason: string
  }
  position?: number
}

export interface ComputeEnvironmentDraft {
  lab_id: string
  environment_key: string
  name: string
  description: string
  runner_protocol_version: "airalogy.compute-runner.v1"
  image_ref: string
  runtime_version: string
  allowed_languages: Array<"python" | "r">
  resource_limits: ComputeResourceLimits
  network_policy: "none" | "egress_allowlist"
  allowed_egress_hosts: string[]
  input_schema: Record<string, unknown>
  result_schema: Record<string, unknown>
  software_manifest: Record<string, unknown>
  estimated_cost_per_hour?: string | null
  currency?: string | null
  risk: "low" | "medium" | "high"
  enabled: boolean
  reason: string
}

export interface ComputeEnvironmentRevisionDraft extends ComputeEnvironmentDraft {
  expected_revision: number
}

export interface ResearchComputePreview {
  preview_digest: string
  command: Record<string, unknown>
  effects: string[]
}

async function requiredData<T>(config: Parameters<typeof request<T>>[0], message: string) {
  const { data, error } = await request<T>(config)
  if (error)
    throw error
  if (!data)
    throw new Error(message)
  return data
}

export function fetchResearchComputeEnvironments(labId: string) {
  return requiredData<{ items: ResearchComputeEnvironment[] }>({
    url: "/research-compute-environments",
    params: { lab_id: labId },
  }, "Research Compute Environment catalog returned no data")
}

export function previewResearchComputeEnvironment(payload: ComputeEnvironmentDraft) {
  return requiredData<ResearchComputePreview>({
    url: "/research-compute-environments/preview",
    method: "post",
    data: payload,
  }, "Research Compute Environment preview returned no data")
}

export function createResearchComputeEnvironment(
  payload: ComputeEnvironmentDraft & { preview_digest: string },
) {
  return requiredData<ResearchComputeEnvironment>({
    url: "/research-compute-environments",
    method: "post",
    data: payload,
  }, "Research Compute Environment creation returned no data")
}

export function previewResearchComputeEnvironmentRevision(
  environmentId: string,
  payload: ComputeEnvironmentRevisionDraft,
) {
  return requiredData<ResearchComputePreview>({
    url: `/research-compute-environments/${environmentId}/revisions/preview`,
    method: "post",
    data: payload,
  }, "Research Compute Environment revision preview returned no data")
}

export function createResearchComputeEnvironmentRevision(
  environmentId: string,
  payload: ComputeEnvironmentRevisionDraft & { preview_digest: string },
) {
  return requiredData<ResearchComputeEnvironment>({
    url: `/research-compute-environments/${environmentId}/revisions`,
    method: "post",
    data: payload,
  }, "Research Compute Environment revision returned no data")
}

File: apps/api/app/services/research_compute.py
"""Versioned, Lab-governed compute-environment catalog helpers."""

from __future__ import annotations

import re
from decimal import Decimal
from typing import Any
from uuid import UUID

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.research_execution import (
    ResearchComputeEnvironment,
    ResearchComputeEnvironmentRevision,
)

COMPUTE_KEY_RE = re.compile(r"^[a-z0-9][a-z0-9._-]{1,127}$")
OCI_DIGEST_RE = re.compile(
    r"^(?![^@]*://)[a-z0-9][a-z0-9._:/-]{0,1975}@sha256:[0-9a-f]{64}$"
)
EGRESS_HOST_RE = re.compile(
    r"^(?:[a-z0-9](?:[a-z0-9.-]{0,251}[a-z0-9])?|\[[0-9a-f:]+\])(?::[1-9][0-9]{0,4})?$"
)


def _money(value: Decimal | None) -> str | None:
    return format(value.normalize(), "f") if value is not None else None


def compute_environment_snapshot(
    environment: ResearchComputeEnvironment,
    revision: ResearchComputeEnvironmentRevision,
) -> dict[str, Any]:
    return {
        "key": f"compute:{environment.id}",
        "version": str(revision.revision),
        "kind": "compute",
        "name": revision.name,
        "description": revision.description,
        "source_type": "research_compute_environment_revision",
        "source_id": str(environment.id),
        "source_revision_id": str(revision.id),
        "executor_types": ["compute_runner"],
        "risk": revision.risk,
        "input_schema": revision.input_schema,
        "output_schema": revision.result_schema,
        "available": bool(revision.enabled and environment.archived_at is None),
        "unavailable_reason": (
            ""
            if revision.enabled and environment.archived_at is None
            else "Compute environment is disabled or archived"
        ),
        "metadata": {
            "lab_id": str(environment.lab_id),
            "environment_key": environment.environment_key,
            "environment_revision": revision.revision,
            "runner_protocol_version": revision.runner_protocol_version,
            "image_ref": revision.image_ref,
            "runtime_version": revision.runtime_version,
            "allowed_languages": revision.allowed_languages,
            "resource_limits": revision.resource_limits,
            "network_policy": revision.network_policy,
            "allowed_egress_hosts": revision.allowed_egress_hosts,
            "software_manifest": revision.software_manifest,
            "estimated_cost_per_hour": _money(revision.estimated_cost_per_hour),
            "currency": revision.currency,
            "change_reason": revision.reason,
        },
    }


async def latest_compute_environment_revision(
    db_session: AsyncSession,
    environment_id: UUID,
    *,
    lock: bool = False,
) -> ResearchComputeEnvironmentRevision | None:
    statement = (
        select(ResearchComputeEnvironmentRevision)
        .where(
            ResearchComputeEnvironmentRevision.compute_environment_id == environment_id
        )
        .order_by(ResearchComputeEnvironmentRevision.revision.desc())
        .limit(1)
    )
    if lock:
        statement = statement.with_for_update()
    return (await db_session.scalars(statement)).first()


async def latest_compute_environment_rows(
    db_session: AsyncSession,
    *,
    lab_id: UUID,
    enabled_only: bool,
) -> list[tuple[ResearchComputeEnvironment, ResearchComputeEnvironmentRevision]]:
    latest = (
        select(
            ResearchComputeEnvironmentRevision.compute_environment_id.label(
                "environment_id"
            ),
            func.max(ResearchComputeEnvironmentRevision.revision).label("revision"),
        )
        .group_by(ResearchComputeEnvironmentRevision.compute_environment_id)
        .subquery()
    )
    statement = (
        select(ResearchComputeEnvironment, ResearchComputeEnvironmentRevision)
        .join(
            latest,
            latest.c.environment_id == ResearchComputeEnvironment.id,
        )
        .join(
            ResearchComputeEnvironmentRevision,
            (
                ResearchComputeEnvironmentRevision.compute_environment_id
                == latest.c.environment_id
            )
            & (ResearchComputeEnvironmentRevision.revision == latest.c.revision),
        )
        .where(
            ResearchComputeEnvironment.lab_id == lab_id,
            ResearchComputeEnvironment.archived_at.is_(None),
        )
        .order_by(
            ResearchComputeEnvironmentRevision.name,
            ResearchComputeEnvironment.id,
        )
    )
    if enabled_only:
        statement = statement.where(
            ResearchComputeEnvironmentRevision.enabled.is_(True)
        )
    return list((await db_session.execute(statement)).all())


async def all_compute_environment_revision_rows(
    db_session: AsyncSession,
    *,
    lab_id: UUID,
    enabled_only: bool,
) -> list[tuple[ResearchComputeEnvironment, ResearchComputeEnvironmentRevision]]:
    """List exact revisions so old Task contracts remain bindable and reproducible."""

    statement = (
        select(ResearchComputeEnvironment, ResearchComputeEnvironmentRevision)
        .join(
            ResearchComputeEnvironmentRevision,
            ResearchComputeEnvironmentRevision.compute_environment_id
            == ResearchComputeEnvironment.id,
        )
        .where(
            ResearchComputeEnvironment.lab_id == lab_id,
            ResearchComputeEnvironment.archived_at.is_(None),
        )
        .order_by(
            ResearchComputeEnvironmentRevision.name,
            ResearchComputeEnvironmentRevision.revision.desc(),
            ResearchComputeEnvironment.id,
        )
    )
    if enabled_only:
        statement = statement.where(
            ResearchComputeEnvironmentRevision.enabled.is_(True)
        )
    return list((await db_session.execute(statement)).all())
